<a href="https://colab.research.google.com/github/sameersahu473/comp1/blob/main/Ahan_mods.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
! pip install kaggle

In [2]:
! mkdir ~/.kaggle

In [3]:
! cp kaggle.json ~/.kaggle/

In [4]:
! chmod 600 ~/.kaggle/kaggle.json


In [5]:
!kaggle competitions download -c cascade-cup-22

  0% 0.00/317k [00:00<?, ?B/s]
100% 317k/317k [00:00<00:00, 21.6MB/s]
 65% 11.0M/16.9M [00:00<00:00, 112MB/s]
100% 16.9M/16.9M [00:00<00:00, 106MB/s]
  0% 0.00/4.31M [00:00<?, ?B/s]
100% 4.31M/4.31M [00:00<00:00, 37.9MB/s]


In [6]:
!unzip /content/train.csv.zip

Archive:  /content/train.csv.zip
  inflating: train.csv               


In [7]:
!unzip /content/test.csv.zip

Archive:  /content/test.csv.zip
  inflating: test.csv                


In [8]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [9]:
df = pd.read_csv('train.csv')
df_test = pd.read_csv('test.csv')

In [10]:
df.head()

order_time  order_id  ... session_time cancelled_time
0  2021-01-26 02:21:35    556753  ...          NaN            NaN
1  2021-01-26 02:33:16    556754  ...     3.266667            NaN
2  2021-01-26 02:39:49    556755  ...     9.816667            NaN
3  2021-01-26 02:47:53    556756  ...    17.533333            NaN
4  2021-01-26 03:06:30    556757  ...     1.350000            NaN

[5 rows x 20 columns]

In [11]:
df_test.head()

order_time  order_id  ... reassigned_order session_time
0  2021-02-06 10:03:24    130231  ...              NaN   273.400000
1  2021-02-06 10:03:26    130232  ...              NaN   252.100000
2  2021-02-06 10:03:27    130233  ...              NaN   241.383333
3  2021-02-06 10:03:29    130234  ...              NaN   291.933333
4  2021-02-06 10:03:35    130235  ...              NaN   247.133333

[5 rows x 16 columns]

In [12]:
df['rider_id'].value_counts()

237      228
190      213
11998    209
170      203
853      200
        ... 
9339       1
20693      1
16603      1
20692      1
4895       1
Name: rider_id, Length: 19537, dtype: int64

In [13]:
df.shape

(450000, 20)

In [14]:
df.isnull().sum()

order_time                   0
order_id                     0
order_date                   0
allot_time                   0
accept_time                157
pickup_time               2421
delivered_time            5218
rider_id                     0
first_mile_distance          0
last_mile_distance           0
alloted_orders           16948
delivered_orders         17341
cancelled                    0
undelivered_orders       17341
lifetime_order_count        53
reassignment_method     436256
reassignment_reason     436247
reassigned_order        436247
session_time              3675
cancelled_time          444782
dtype: int64

In [15]:
df = df.drop(columns=['order_id','pickup_time','delivered_time'],axis =1)

In [16]:
df.head()

order_time           order_date  ... session_time cancelled_time
0  2021-01-26 02:21:35  2021-01-26 00:00:00  ...          NaN            NaN
1  2021-01-26 02:33:16  2021-01-26 00:00:00  ...     3.266667            NaN
2  2021-01-26 02:39:49  2021-01-26 00:00:00  ...     9.816667            NaN
3  2021-01-26 02:47:53  2021-01-26 00:00:00  ...    17.533333            NaN
4  2021-01-26 03:06:30  2021-01-26 00:00:00  ...     1.350000            NaN

[5 rows x 17 columns]

In [17]:
df_test = df_test.drop(['order_id'],axis=1)

In [18]:
df.isnull().sum()

order_time                   0
order_date                   0
allot_time                   0
accept_time                157
rider_id                     0
first_mile_distance          0
last_mile_distance           0
alloted_orders           16948
delivered_orders         17341
cancelled                    0
undelivered_orders       17341
lifetime_order_count        53
reassignment_method     436256
reassignment_reason     436247
reassigned_order        436247
session_time              3675
cancelled_time          444782
dtype: int64

In [19]:
df['cancelled'].value_counts()

0    444782
1      5218
Name: cancelled, dtype: int64

In [20]:
for col in df.columns:  
    dataTypeSeries = df[col].dtypes
    print('Data type of '+ col+ ' column of Dataframe :')
    print(dataTypeSeries)

Data type of order_time column of Dataframe :
object
Data type of order_date column of Dataframe :
object
Data type of allot_time column of Dataframe :
object
Data type of accept_time column of Dataframe :
object
Data type of rider_id column of Dataframe :
int64
Data type of first_mile_distance column of Dataframe :
float64
Data type of last_mile_distance column of Dataframe :
float64
Data type of alloted_orders column of Dataframe :
float64
Data type of delivered_orders column of Dataframe :
float64
Data type of cancelled column of Dataframe :
int64
Data type of undelivered_orders column of Dataframe :
float64
Data type of lifetime_order_count column of Dataframe :
float64
Data type of reassignment_method column of Dataframe :
object
Data type of reassignment_reason column of Dataframe :
object
Data type of reassigned_order column of Dataframe :
float64
Data type of session_time column of Dataframe :
float64
Data type of cancelled_time column of Dataframe :
object


In [21]:
#creating the datetime column
df['allot_time'] = pd.to_datetime(df['allot_time'].str.strip(), format='%Y/%m/%d %H:%M:%S')
df_test['allot_time'] = pd.to_datetime(df_test['allot_time'].str.strip(), format='%Y/%m/%d %H:%M:%S')

In [22]:
df['accept_time'] = pd.to_datetime(df['accept_time'].str.strip(), format='%Y/%m/%d %H:%M:%S')
df_test['accept_time'] = pd.to_datetime(df_test['accept_time'].str.strip(), format='%Y/%m/%d %H:%M:%S')

In [23]:
for col in df.columns:  
    dataTypeSeries = df[col].dtypes
    print('Data type of '+ col+ ' column of Dataframe :')
    print(dataTypeSeries)

Data type of order_time column of Dataframe :
object
Data type of order_date column of Dataframe :
object
Data type of allot_time column of Dataframe :
datetime64[ns]
Data type of accept_time column of Dataframe :
datetime64[ns]
Data type of rider_id column of Dataframe :
int64
Data type of first_mile_distance column of Dataframe :
float64
Data type of last_mile_distance column of Dataframe :
float64
Data type of alloted_orders column of Dataframe :
float64
Data type of delivered_orders column of Dataframe :
float64
Data type of cancelled column of Dataframe :
int64
Data type of undelivered_orders column of Dataframe :
float64
Data type of lifetime_order_count column of Dataframe :
float64
Data type of reassignment_method column of Dataframe :
object
Data type of reassignment_reason column of Dataframe :
object
Data type of reassigned_order column of Dataframe :
float64
Data type of session_time column of Dataframe :
float64
Data type of cancelled_time column of Dataframe :
object


In [24]:
df['accept_to_allot'] = (df['accept_time'] - df['allot_time'])
df_test['accept_to_allot'] = (df_test['accept_time'] - df_test['allot_time'])

In [25]:
df.head()

order_time           order_date  ... cancelled_time accept_to_allot
0  2021-01-26 02:21:35  2021-01-26 00:00:00  ...            NaN 0 days 00:00:09
1  2021-01-26 02:33:16  2021-01-26 00:00:00  ...            NaN 0 days 00:00:48
2  2021-01-26 02:39:49  2021-01-26 00:00:00  ...            NaN 0 days 00:00:16
3  2021-01-26 02:47:53  2021-01-26 00:00:00  ...            NaN 0 days 00:00:41
4  2021-01-26 03:06:30  2021-01-26 00:00:00  ...            NaN 0 days 00:00:36

[5 rows x 18 columns]

In [26]:
for col in df.columns:  
    dataTypeSeries = df[col].dtypes
    print('Data type of '+ col+ ' column of Dataframe :')
    print(dataTypeSeries)

Data type of order_time column of Dataframe :
object
Data type of order_date column of Dataframe :
object
Data type of allot_time column of Dataframe :
datetime64[ns]
Data type of accept_time column of Dataframe :
datetime64[ns]
Data type of rider_id column of Dataframe :
int64
Data type of first_mile_distance column of Dataframe :
float64
Data type of last_mile_distance column of Dataframe :
float64
Data type of alloted_orders column of Dataframe :
float64
Data type of delivered_orders column of Dataframe :
float64
Data type of cancelled column of Dataframe :
int64
Data type of undelivered_orders column of Dataframe :
float64
Data type of lifetime_order_count column of Dataframe :
float64
Data type of reassignment_method column of Dataframe :
object
Data type of reassignment_reason column of Dataframe :
object
Data type of reassigned_order column of Dataframe :
float64
Data type of session_time column of Dataframe :
float64
Data type of cancelled_time column of Dataframe :
object
Data

In [27]:
df['accept_sec'] = df[:5]['accept_to_allot'].astype('timedelta64[s]')
df_test['accept_sec'] = df_test[:5]['accept_to_allot'].astype('timedelta64[s]')

In [28]:
df.head()

order_time           order_date  ... accept_to_allot accept_sec
0  2021-01-26 02:21:35  2021-01-26 00:00:00  ... 0 days 00:00:09        9.0
1  2021-01-26 02:33:16  2021-01-26 00:00:00  ... 0 days 00:00:48       48.0
2  2021-01-26 02:39:49  2021-01-26 00:00:00  ... 0 days 00:00:16       16.0
3  2021-01-26 02:47:53  2021-01-26 00:00:00  ... 0 days 00:00:41       41.0
4  2021-01-26 03:06:30  2021-01-26 00:00:00  ... 0 days 00:00:36       36.0

[5 rows x 19 columns]

In [29]:
df_test.head()

order_time           order_date  ... accept_to_allot accept_sec
0  2021-02-06 10:03:24  2021-02-06 00:00:00  ... 0 days 00:00:26       26.0
1  2021-02-06 10:03:26  2021-02-06 00:00:00  ... 0 days 00:00:09        9.0
2  2021-02-06 10:03:27  2021-02-06 00:00:00  ... 0 days 00:01:20       80.0
3  2021-02-06 10:03:29  2021-02-06 00:00:00  ... 0 days 00:00:23       23.0
4  2021-02-06 10:03:35  2021-02-06 00:00:00  ... 0 days 00:01:00       60.0

[5 rows x 17 columns]

In [30]:
#creating the datetime column
df['order_time'] = pd.to_datetime(df['order_time'].str.strip(), format='%Y/%m/%d %H:%M:%S')
df_test['order_time'] = pd.to_datetime(df_test['order_time'].str.strip(), format='%Y/%m/%d %H:%M:%S')

In [31]:
df['order_to_allot'] = (df['accept_time'] - df['order_time'])
df_test['order_to_allot'] = (df_test['accept_time'] - df_test['order_time'])

In [32]:
df['order_sec'] = df[:5]['order_to_allot'].astype('timedelta64[s]')
df_test['order_sec'] = df_test[:5]['order_to_allot'].astype('timedelta64[s]')

In [33]:
df.head()

order_time           order_date  ...  order_to_allot order_sec
0 2021-01-26 02:21:35  2021-01-26 00:00:00  ... 0 days 00:00:33      33.0
1 2021-01-26 02:33:16  2021-01-26 00:00:00  ... 0 days 00:01:29      89.0
2 2021-01-26 02:39:49  2021-01-26 00:00:00  ... 0 days 00:00:24      24.0
3 2021-01-26 02:47:53  2021-01-26 00:00:00  ... 0 days 00:01:13      73.0
4 2021-01-26 03:06:30  2021-01-26 00:00:00  ... 0 days 00:01:27      87.0

[5 rows x 21 columns]

In [34]:
df['allot_sec'] = df['order_sec'] - df['accept_sec']
df_test['allot_sec'] = df_test['order_sec'] - df_test['accept_sec']

In [35]:
df_test = df_test.drop(['order_time','allot_time','accept_time','accept_to_allot','order_to_allot'],axis=1)
df= df.drop(['order_time','allot_time','accept_time','accept_to_allot','order_to_allot'],axis=1)

In [36]:
df['order_date'] = pd.to_datetime(df['order_date'].str.strip(), format='%Y/%m/%d %H:%M:%S')
df_test['order_date'] = pd.to_datetime(df_test['order_date'].str.strip(), format='%Y/%m/%d %H:%M:%S')

In [37]:
df['year'] = pd.DatetimeIndex(df['order_date']).year
df['month'] = pd.DatetimeIndex(df['order_date']).month


df_test['year'] = pd.DatetimeIndex(df_test['order_date']).year
df_test['month'] = pd.DatetimeIndex(df_test['order_date']).month

In [38]:
df.head()

order_date  rider_id  first_mile_distance  ...  allot_sec  year  month
0 2021-01-26     11696               1.5666  ...       24.0  2021      1
1 2021-01-26     18117               2.5207  ...       41.0  2021      1
2 2021-01-26     18623               2.2074  ...        8.0  2021      1
3 2021-01-26     15945               2.1894  ...       32.0  2021      1
4 2021-01-26     17589               2.7870  ...       51.0  2021      1

[5 rows x 19 columns]

In [39]:
df.columns

Index(['order_date', 'rider_id', 'first_mile_distance', 'last_mile_distance',
       'alloted_orders', 'delivered_orders', 'cancelled', 'undelivered_orders',
       'lifetime_order_count', 'reassignment_method', 'reassignment_reason',
       'reassigned_order', 'session_time', 'cancelled_time', 'accept_sec',
       'order_sec', 'allot_sec', 'year', 'month'],
      dtype='object')

In [40]:
df['reassignment_reason'].value_counts()

Auto Reassignment basis Inaction. coreengine.tasks.repush_order_to_aa_bucket    7212
Reassignment Request from SE portal.                                            5300
Reassign                                                                        1241
Name: reassignment_reason, dtype: int64

In [41]:
df['reassignment_method'].value_counts()

auto      13383
manual      361
Name: reassignment_method, dtype: int64

In [42]:
df.dtypes

order_date              datetime64[ns]
rider_id                         int64
first_mile_distance            float64
last_mile_distance             float64
alloted_orders                 float64
delivered_orders               float64
cancelled                        int64
undelivered_orders             float64
lifetime_order_count           float64
reassignment_method             object
reassignment_reason             object
reassigned_order               float64
session_time                   float64
cancelled_time                  object
accept_sec                     float64
order_sec                      float64
allot_sec                      float64
year                             int64
month                            int64
dtype: object

In [43]:
num_col = [c for i, c in enumerate(df.columns) if df.dtypes[i] in [np.int, np.float]]
len(num_col)

15

In [44]:
num_col = [c for i, c in enumerate(df.columns) if df.dtypes[i] in [np.object]]
len(num_col)

3

In [45]:
df.isnull().sum()

order_date                   0
rider_id                     0
first_mile_distance          0
last_mile_distance           0
alloted_orders           16948
delivered_orders         17341
cancelled                    0
undelivered_orders       17341
lifetime_order_count        53
reassignment_method     436256
reassignment_reason     436247
reassigned_order        436247
session_time              3675
cancelled_time          444782
accept_sec              449995
order_sec               449995
allot_sec               449995
year                         0
month                        0
dtype: int64

In [46]:
df['reassignment_reason'] = df['reassignment_reason'].replace(np.nan, 0)

In [47]:
df['reassignment_reason_freq'] = df['reassignment_reason'].replace(('Auto Reassignment basis Inaction. coreengine.tasks.repush_order_to_aa_bucket','Reassignment Request from SE portal.','Reassign','0'),(7212,5300,1241,436256))

In [48]:
df['reassignment_reason'].value_counts()

0                                                                               436247
Auto Reassignment basis Inaction. coreengine.tasks.repush_order_to_aa_bucket      7212
Reassignment Request from SE portal.                                              5300
Reassign                                                                          1241
Name: reassignment_reason, dtype: int64

In [49]:
pip install --upgrade category_encoders

     |████████████████████████████████| 82 kB 244 kB/s 


In [50]:
from category_encoders import TargetEncoder

encoder = TargetEncoder()
df['reassignment_reason']= encoder.fit_transform(df['reassignment_reason'], df['cancelled'])

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [51]:
df.columns

Index(['order_date', 'rider_id', 'first_mile_distance', 'last_mile_distance',
       'alloted_orders', 'delivered_orders', 'cancelled', 'undelivered_orders',
       'lifetime_order_count', 'reassignment_method', 'reassignment_reason',
       'reassigned_order', 'session_time', 'cancelled_time', 'accept_sec',
       'order_sec', 'allot_sec', 'year', 'month', 'reassignment_reason_freq'],
      dtype='object')

In [52]:
df['reassignment_method'] = df['reassignment_method'].replace(np.nan, 0)

In [53]:
df['reassignment_method'].value_counts()

0         436256
auto       13383
manual       361
Name: reassignment_method, dtype: int64

In [54]:
df['reassignment_method_freq'] = df['reassignment_reason'].replace(('auto','manual','0'),(13383,361,436256))

In [55]:
encoder = TargetEncoder()
df['reassignment_method']= encoder.fit_transform(df['reassignment_method'], df['cancelled'])

In [56]:
df['total_distance'] = df['first_mile_distance'] + df['last_mile_distance']

In [57]:
df.columns

Index(['order_date', 'rider_id', 'first_mile_distance', 'last_mile_distance',
       'alloted_orders', 'delivered_orders', 'cancelled', 'undelivered_orders',
       'lifetime_order_count', 'reassignment_method', 'reassignment_reason',
       'reassigned_order', 'session_time', 'cancelled_time', 'accept_sec',
       'order_sec', 'allot_sec', 'year', 'month', 'reassignment_reason_freq',
       'reassignment_method_freq', 'total_distance'],
      dtype='object')

In [58]:
df = df.drop(columns=['reassignment_method1','reassignment_reason1','cancelled_time'],axis =1)

KeyError: ignored

In [59]:
df.head()

order_date  rider_id  ...  reassignment_method_freq  total_distance
0 2021-01-26     11696  ...                  0.010306          4.2166
1 2021-01-26     18117  ...                  0.010306          5.2807
2 2021-01-26     18623  ...                  0.010306          7.0074
3 2021-01-26     15945  ...                  0.010306          8.5694
4 2021-01-26     17589  ...                  0.010306          6.7970

[5 rows x 22 columns]

In [60]:
df_test.head()

order_date  rider_id  first_mile_distance  ...  allot_sec  year  month
0 2021-02-06     12884               1.6585  ...       25.0  2021      2
1 2021-02-06      3541               2.0709  ...        1.0  2021      2
2 2021-02-06       603               1.3884  ...       47.0  2021      2
3 2021-02-06      3414               1.9039  ...        1.0  2021      2
4 2021-02-06      1426               0.8275  ...        8.0  2021      2

[5 rows x 17 columns]

There were a total of 17341 missing values in both the **'delivered_orders'** and **'undelivered_orders'** column as found above i.e a total of 450000 - 17341 = **432659 valid values** in these two colums.

In [70]:
rows = 450000
count = 0
for i in range(rows):
  if df['delivered_orders'][i] + df['undelivered_orders'][i] == df['alloted_orders'][i]:
    count += 1
print(count)

432659


Clearly for all the valid rows it can be seen that the sum of **delivered_orders** and **undelivered_orders** is equal to **alloted_orders** so we can replace the two columns by one column.

In [74]:
df = df[pd.notnull(df['delivered_orders'])]

In [75]:
df.shape

(432659, 22)

In [76]:
df.isnull().sum()

order_date                       0
rider_id                         0
first_mile_distance              0
last_mile_distance               0
alloted_orders                   0
delivered_orders                 0
cancelled                        0
undelivered_orders               0
lifetime_order_count             0
reassignment_method              0
reassignment_reason              0
reassigned_order            419393
session_time                  3344
cancelled_time              428296
accept_sec                  432654
order_sec                   432654
allot_sec                   432654
year                             0
month                            0
reassignment_reason_freq         0
reassignment_method_freq         0
total_distance                   0
dtype: int64

Hence all the rows with missing values in both the **'delivered_orders'** and **'undelivered_orders'** column have been removed.We run the same on the test data.

In [79]:
df_test = df_test[pd.notnull(df_test['delivered_orders'])]

In [80]:
df_test.shape

(139960, 17)

In [81]:
df_test.isnull().sum()

order_date                   0
rider_id                     0
first_mile_distance          0
last_mile_distance           0
alloted_orders               0
delivered_orders             0
undelivered_orders           0
lifetime_order_count         0
reassignment_method     135485
reassignment_reason     135482
reassigned_order        135482
session_time              4786
accept_sec              139955
order_sec               139955
allot_sec               139955
year                         0
month                        0
dtype: int64

We drop the **delivered_orders** and **undelivered_orders** columns and replace them with **delivery_success** column as shown.

In [84]:
df['delivery_success'] = (df['delivered_orders']/df['alloted_orders'])
df_test['delivery_success'] = (df_test['delivered_orders']/df_test['alloted_orders'])

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [87]:
df = df.drop(['delivered_orders', 'undelivered_orders'], axis=1)
df_test = df_test.drop(['delivered_orders', 'undelivered_orders'], axis=1)

In [88]:
df.head()

order_date  rider_id  ...  total_distance  delivery_success
0 2021-01-26     11696  ...          4.2166               1.0
1 2021-01-26     18117  ...          5.2807               1.0
2 2021-01-26     18623  ...          7.0074               1.0
3 2021-01-26     15945  ...          8.5694               1.0
4 2021-01-26     17589  ...          6.7970               1.0

[5 rows x 21 columns]